Adaugare optimizare rute si afisare traseu cu ordinea punctelor pe harta

MagdaIonescu opened the pull request:
Am implementat calculul rutei optimizate, afisarea traseelor folosind Leaflet, marcarea punctelor cu ordinea de vizitare, adaugarea legendei pe harta, stilizarea paginilor (navbar, footer, formulare, tabele, harta).

MagdaIonescu committed:
Adaugare optimizare rute si afisare traseu cu ordinea punctelor pe harta

Am implementat calculul rutei optimizate, afisarea traseelor neoptimizat/optimizat folosind Leaflet, marcarea punctelor cu ordinea de vizitare, adaugarea legendei pe harta, stilizarea paginilor (navbar, footer, formulare, tabele, harta).

pdany1116 commented on colectare-date/colectare-date/Controllers/HartaController.cs:
```diff
@@ -18,7 +21,148 @@ public IActionResult Harta()
+            var distantaTraseuNeoptimizat = CalculeazaDistanta(rutaNeoptimizata, matrice);
+            var distantaTraseuOptimizat = CalculeazaDistanta(rutaOptimizata, matrice);
+
+            var timpEstimativ = rutaOptimizata.Count * 1;
```
Ati respectat cerinta in care presupunem ca timpul de colectare e 1 min, dar timpul de deplasare nu l-ati luat in calcul

MagdaIonescu replied:
Multumim pentru sugestii, le vom lua in calcul si vom imbunatati forma actuala a aplicatiei. 

Intr-adevar, am calculat timpul de colectare al pubelelor, de fapt, nu timpul complet pentru ca nu am stiut cum sa procedam pentru viteza masinii.  Vom aproxima viteza masinii cu o anumita valoare (poate 20 sau 30 de km)?

pdany1116 replied:
Este o solutie. Aveti alte idei pe langa a aproxima viteza?

MagdaIonescu replied:
Avand in vedere ca nu avem matricea cu timpii, nu stim cum altfel am putea face.
O alta idee ar fi fost aceasta, sa ne generam si matricea care sa contina timpii pentru fiecare distanta, sa o incarcam si sa o folosim pentru a calcula timpul estimat = timp deplasare + timp colectare..

pdany1116 replied:
Dap, bravo, asta ar fi solutia, sa folosim matricea cu timpii. Nu e obligatoriu de implementat, ramana ca challenge daca doriti.

pdany1116 commented on colectare-date/colectare-date/Controllers/HartaController.cs:
```diff
@@ -18,7 +21,148 @@ public IActionResult Harta()
+            var timpEstimativ = rutaOptimizata.Count * 1;
+
+            var rutaFaraStartStop = rutaOptimizata.Where(index => index > 0 && index <= colectari.Count).ToList();
+            var colectariOrdonate = rutaFaraStartStop
```
Variabila nefolosita

pdany1116 commented on colectare-date/colectare-date/Controllers/HartaController.cs:
```diff
@@ -18,7 +21,148 @@ public IActionResult Harta()
+            return View(model);
+        }
+
+        public int[,] IncarcaMatrice()
```
Metoda asta ar trebui sa fie privata (+ celelalte de mai jos care nu sunt rute), e folosita doar in ruta principala servita de metoda Harta

pdany1116 commented on colectare-date/colectare-date/Controllers/HartaController.cs:
```diff
@@ -18,7 +21,148 @@ public IActionResult Harta()
+                .Select(index => colectari[index - 1])
+                .ToList();
+
+            var model = new RutaOptimizataViewModel
```
Foarte bine ca am facut o clasa RutaOptimizataViewModel, dar o folosim doar aici, in ruta RutaOptimizata folosim partial clasa in combinatie cu ViewBag

pdany1116 commented on colectare-date/colectare-date/Views/Harta/Harta.cshtml:
```diff
@@ -1,50 +1,133 @@
+</div>
 
 <link rel="stylesheet" href="https://unpkg.com/leaflet@1.9.4/dist/leaflet.css" />
+<link rel="stylesheet" href="https://unpkg.com/leaflet-routing-machine@3.2.12/dist/leaflet-routing-machine.css" />
```
Bravo!

pdany1116 commented on colectare-date/colectare-date/Controllers/HartaController.cs:
```diff
@@ -18,7 +21,148 @@ public IActionResult Harta()
+            var matrice = IncarcaMatrice();
+            var rutaOptimizata = CalculeazaRutaOptimizata(matrice);
+
+            var distantaTraseuNeoptimizat = CalculeazaDistanta(rutaNeoptimizata, matrice);
+            var distantaTraseuOptimizat = CalculeazaDistanta(rutaOptimizata, matrice);
```
La distante nu adaugam distanta de la punctul de start pana la primul punct din ruta si nici distanta de la ultimul punct din ruta pana la punctul de stop. Pierdem astfel cativa km importanti.

pdany1116 reviewed (commented):
Buna treaba, felicitari! Am lasat cateva sugestii si intrebari

pdany1116 commented on colectare-date/colectare-date/Views/Harta/Harta.cshtml:
```diff
@@ -1,50 +1,133 @@
+    <div class="rezultate-container">
+        <h2>Rezultate Optimizare</h2>
+        <ul>
+            <li><strong>Distanta traseu neoptimizat:</strong> @Model.DistantaTraseuNeoptimizat metri</li>
+            <li><strong>Distanta traseu optimizat:</strong> @Model.DistantaTraseuOptimizat metri</li>
+            <li><strong>Diferenta de distanta parcursa dintre traseul parcurs de masina si noul traseu optimizat:</strong> @Model.DiferentaDistanta metri</li>
```
Cred ca are mai mult sens sa fie in km decat in metri

MagdaIonescu committed:
Optimizare-Rute - realizare sugestii

Am calculat timpul estimativ, folosind viteza medie de 30 km/h. Am facut metodele private, am eliminat combinatia dintre RutaOptimizata si ViewBag, folosind RutaOptimizataViewModel.

Co-Authored-By: EduardMunteanu <127775254+EduardMunteanu@users.noreply.github.com>
